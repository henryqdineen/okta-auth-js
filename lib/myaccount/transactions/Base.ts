import { OktaAuthInterface } from '../../types';
import { TransactionLink } from '../request';

type TransactionOptions = {
  // TODO: move res type to http module
  res: {
    headers: Record<string, string>;
    _http: Record<string, string | object>;
    _links?: Record<string, TransactionLink>;
    [property: string]: unknown;
  };
  accessToken: string;
};
export default class BaseTransaction {
  // Deprecated
  headers?: Record<string, string>;
  _http: Record<string, string | object>;

  constructor(oktaAuth: OktaAuthInterface, options: TransactionOptions) {
    const { res } = options;
    const { headers, _http, ...rest } = res;
    
    // assign required fields from res
    if (headers) {
      this.headers = headers;
    }
    this._http = _http;

    // add all rest fields from res
    Object.keys(rest).forEach(key => {
      if (key === '_links') {
        return;
      }
      this[key] = rest[key];
    });
  }
}
