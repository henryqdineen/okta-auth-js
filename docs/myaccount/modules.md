[@okta/okta-auth-js/myaccount](README.md) / Modules

# @okta/okta-auth-js/myaccount

## Table of contents

### Modules

- [index](modules/index.md)
- [types](modules/types.md)
