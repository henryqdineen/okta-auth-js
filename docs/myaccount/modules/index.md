[@okta/okta-auth-js/myaccount](../README.md) / [Modules](../modules.md) / index

# Module: index

## Table of contents

### Functions

- [addEmail](index.md#addemail)
- [addPhone](index.md#addphone)
- [deleteEmail](index.md#deleteemail)
- [deletePhone](index.md#deletephone)
- [getEmail](index.md#getemail)
- [getEmailChallenge](index.md#getemailchallenge)
- [getEmails](index.md#getemails)
- [getPhone](index.md#getphone)
- [getPhones](index.md#getphones)
- [getProfile](index.md#getprofile)
- [getProfileSchema](index.md#getprofileschema)
- [sendEmailChallenge](index.md#sendemailchallenge)
- [sendPhoneChallenge](index.md#sendphonechallenge)
- [updateProfile](index.md#updateprofile)
- [verifyEmailChallenge](index.md#verifyemailchallenge)
- [verifyPhoneChallenge](index.md#verifyphonechallenge)

## Functions

### addEmail

▸ **addEmail**(`oktaAuth`, `options?`): `Promise`<[`EmailTransaction`](../classes/types.EmailTransaction.md)\>

**`scope:`** okta.myaccount.email.manage

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`EmailTransaction`](../classes/types.EmailTransaction.md)\>

#### Defined in

[emailApi.ts:45](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/emailApi.ts#L45)

___

### addPhone

▸ **addPhone**(`oktaAuth`, `options?`): `Promise`<[`PhoneTransaction`](../classes/types.PhoneTransaction.md)\>

**`scope:`** okta.myaccount.phone.manage

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`PhoneTransaction`](../classes/types.PhoneTransaction.md)\>

#### Defined in

[phoneApi.ts:44](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/phoneApi.ts#L44)

___

### deleteEmail

▸ **deleteEmail**(`oktaAuth`, `options?`): `Promise`<[`BaseTransaction`](../classes/types.BaseTransaction.md)\>

**`scope:`** okta.myaccount.email.manage

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`BaseTransaction`](../classes/types.BaseTransaction.md)\>

#### Defined in

[emailApi.ts:63](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/emailApi.ts#L63)

___

### deletePhone

▸ **deletePhone**(`oktaAuth`, `options?`): `Promise`<[`BaseTransaction`](../classes/types.BaseTransaction.md)\>

**`scope:`** okta.myaccount.phone.manage

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`BaseTransaction`](../classes/types.BaseTransaction.md)\>

#### Defined in

[phoneApi.ts:62](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/phoneApi.ts#L62)

___

### getEmail

▸ **getEmail**(`oktaAuth`, `options?`): `Promise`<[`EmailTransaction`](../classes/types.EmailTransaction.md)\>

**`scope:`** okta.myaccount.email.read

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`EmailTransaction`](../classes/types.EmailTransaction.md)\>

#### Defined in

[emailApi.ts:28](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/emailApi.ts#L28)

___

### getEmailChallenge

▸ **getEmailChallenge**(`oktaAuth`, `options?`): `Promise`<[`EmailChallengeTransaction`](../classes/types.EmailChallengeTransaction.md)\>

**`scope:`** okta.myaccount.email.read

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`EmailChallengeTransaction`](../classes/types.EmailChallengeTransaction.md)\>

#### Defined in

[emailApi.ts:96](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/emailApi.ts#L96)

___

### getEmails

▸ **getEmails**(`oktaAuth`, `options?`): `Promise`<[`EmailTransaction`](../classes/types.EmailTransaction.md)[]\>

**`scope:`** okta.myaccount.email.read

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`EmailTransaction`](../classes/types.EmailTransaction.md)[]\>

#### Defined in

[emailApi.ts:12](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/emailApi.ts#L12)

___

### getPhone

▸ **getPhone**(`oktaAuth`, `options?`): `Promise`<[`PhoneTransaction`](../classes/types.PhoneTransaction.md)\>

**`scope:`** okta.myaccount.phone.read

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`PhoneTransaction`](../classes/types.PhoneTransaction.md)\>

#### Defined in

[phoneApi.ts:27](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/phoneApi.ts#L27)

___

### getPhones

▸ **getPhones**(`oktaAuth`, `options?`): `Promise`<[`PhoneTransaction`](../classes/types.PhoneTransaction.md)[]\>

**`scope:`** okta.myaccount.phone.read

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`PhoneTransaction`](../classes/types.PhoneTransaction.md)[]\>

#### Defined in

[phoneApi.ts:11](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/phoneApi.ts#L11)

___

### getProfile

▸ **getProfile**(`oktaAuth`, `options?`): `Promise`<[`ProfileTransaction`](../classes/types.ProfileTransaction.md)\>

**`scope:`** okta.myaccount.read

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`ProfileTransaction`](../classes/types.ProfileTransaction.md)\>

#### Defined in

[profileApi.ts:11](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/profileApi.ts#L11)

___

### getProfileSchema

▸ **getProfileSchema**(`oktaAuth`, `options?`): `Promise`<[`ProfileSchemaTransaction`](../classes/types.ProfileSchemaTransaction.md)\>

**`scope:`** okta.myaccount.read

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`ProfileSchemaTransaction`](../classes/types.ProfileSchemaTransaction.md)\>

#### Defined in

[profileApi.ts:42](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/profileApi.ts#L42)

___

### sendEmailChallenge

▸ **sendEmailChallenge**(`oktaAuth`, `options?`): `Promise`<[`EmailChallengeTransaction`](../classes/types.EmailChallengeTransaction.md)\>

**`scope:`** okta.myaccount.email.read

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`EmailChallengeTransaction`](../classes/types.EmailChallengeTransaction.md)\>

#### Defined in

[emailApi.ts:79](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/emailApi.ts#L79)

___

### sendPhoneChallenge

▸ **sendPhoneChallenge**(`oktaAuth`, `options?`): `Promise`<[`BaseTransaction`](../classes/types.BaseTransaction.md)\>

**`scope:`** okta.myaccount.phone.manage

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`BaseTransaction`](../classes/types.BaseTransaction.md)\>

#### Defined in

[phoneApi.ts:78](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/phoneApi.ts#L78)

___

### updateProfile

▸ **updateProfile**(`oktaAuth`, `options?`): `Promise`<[`ProfileTransaction`](../classes/types.ProfileTransaction.md)\>

**`scope:`** okta.myaccount.manage

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`ProfileTransaction`](../classes/types.ProfileTransaction.md)\>

#### Defined in

[profileApi.ts:24](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/profileApi.ts#L24)

___

### verifyEmailChallenge

▸ **verifyEmailChallenge**(`oktaAuth`, `options?`): `Promise`<[`BaseTransaction`](../classes/types.BaseTransaction.md)\>

**`scope:`** okta.myaccount.email.manage

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`BaseTransaction`](../classes/types.BaseTransaction.md)\>

#### Defined in

[emailApi.ts:113](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/emailApi.ts#L113)

___

### verifyPhoneChallenge

▸ **verifyPhoneChallenge**(`oktaAuth`, `options?`): `Promise`<[`BaseTransaction`](../classes/types.BaseTransaction.md)\>

**`scope:`** okta.myaccount.phone.manage

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`BaseTransaction`](../classes/types.BaseTransaction.md)\>

#### Defined in

[phoneApi.ts:95](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/phoneApi.ts#L95)
