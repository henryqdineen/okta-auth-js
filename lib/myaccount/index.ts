export * from './profileApi';
export * from './emailApi';
export * from './phoneApi';
