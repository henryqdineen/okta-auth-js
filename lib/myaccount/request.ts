import { 
  BaseTransaction,
  EmailTransaction,
  EmailStatusTransaction,
  EmailChallengeTransaction,
  ProfileTransaction,
  ProfileSchemaTransaction,
  PhoneTransaction
} from './transactions';
import { httpRequest } from '../http';
import { AuthApiError, AuthSdkError } from '../errors';
import { MyAccountRequestOptions as RequestOptions } from './types';
import { OktaAuthInterface } from '../types';

export type TransactionLink = {
  href: string;
  hints?: {
    allow?: string[];
  };
}

type TransactionLinks = {
  self: TransactionLink;
  [property: string]: TransactionLink;
}

type SendRequestOptions = RequestOptions & {
  url: string;
  method: string;
  transactionClassName?: string;
}

type InsufficientAuthenticationError = {
  error: string;
  // eslint-disable-next-line camelcase
  error_description: string;
  // eslint-disable-next-line camelcase
  max_age: string;
}

const parseInsufficientAuthenticationError = (
  header: string
): InsufficientAuthenticationError => {
  if (!header) {
    throw new AuthSdkError('Missing header string');
  }

  return header
    .split(',')
    .map(part => part.trim())
    .map(part => part.split('='))
    .reduce((acc, curr) => {
      // unwrap quotes from value
      acc[curr[0]] = curr[1].replace(/^"(.*)"$/, '$1');
      return acc;
    }, {}) as InsufficientAuthenticationError;
};

/* eslint-disable complexity */
export async function sendRequest<T extends BaseTransaction> (
  oktaAuth: OktaAuthInterface, 
  options: SendRequestOptions
): Promise<T | T[]> {
  const { 
    accessToken: accessTokenObj,
    idToken: idTokenObj 
  } = oktaAuth.tokenManager.getTokensSync();
  
  const idToken = idTokenObj?.idToken;
  const accessToken = options.accessToken || accessTokenObj?.accessToken;
  const { issuer } = oktaAuth.options;
  const { url, method, payload } = options;
  const requestUrl = url.startsWith(issuer!) ? url : `${issuer}${url}`;

  if (!accessToken) {
    throw new AuthSdkError('AccessToken is required to request MyAccount API endpoints.');
  }
  
  let res;
  try {
    res = await httpRequest(oktaAuth, {
      headers: { 'Accept': '*/*;okta-version=1.0.0' },
      accessToken,
      url: requestUrl,
      method,
      ...(payload && { args: payload })
    });
  } catch (err) {
    const errorResp = (err as AuthApiError).xhr;
    if (idToken && errorResp?.status === 403 && !!errorResp?.headers?.['www-authenticate']) {
      const { 
        error, 
        // eslint-disable-next-line camelcase
        error_description,
        // eslint-disable-next-line camelcase
        max_age 
      } = parseInsufficientAuthenticationError(errorResp?.headers?.['www-authenticate']);
      if (error === 'insufficient_authentication_context') {
        const insufficientAuthenticationError = new AuthApiError(
          { 
            errorSummary: error,
            // eslint-disable-next-line camelcase
            errorCauses: [{ errorSummary: error_description }]
          }, 
          errorResp, 
          // eslint-disable-next-line camelcase
          { max_age: +max_age }
        );
        throw insufficientAuthenticationError;
      } else {
        throw err;
      }
    } else {
      throw err;
    }
  }

  const map = {
    EmailTransaction,
    EmailStatusTransaction,
    EmailChallengeTransaction,
    ProfileTransaction,
    ProfileSchemaTransaction,
    PhoneTransaction
  };
  const TransactionClass = map[options.transactionClassName!] || BaseTransaction;

  if (Array.isArray(res)) {
    return res.map(item => new TransactionClass(oktaAuth, { 
      res: item, 
      accessToken
    }));
  }

  return new TransactionClass(oktaAuth, { 
    res, 
    accessToken
  });
}
/* eslint-enable complexity */

export type GenerateRequestFnFromLinksOptions = {
  oktaAuth: OktaAuthInterface;
  accessToken: string;
  methodName: string;
  links: TransactionLinks;
  transactionClassName?: string;
}

type IRequestFnFromLinks = <T extends BaseTransaction>(payload?) => Promise<T | T[]>;

export function generateRequestFnFromLinks ({
  oktaAuth, 
  accessToken,
  methodName,
  links,
  transactionClassName
}: GenerateRequestFnFromLinksOptions): IRequestFnFromLinks {
  for (const method of ['GET', 'POST', 'PUT', 'DELETE']) {
    if (method.toLowerCase() === methodName) {
      const link = links.self;
      return (async (payload?) => sendRequest(oktaAuth, {
        accessToken,
        url: link.href,
        method,
        payload,
        transactionClassName
      }));
    }
  }
  
  const link = links[methodName];
  if (!link) {
    throw new AuthSdkError(`No link is found with methodName: ${methodName}`);
  }

  return (async (payload?) => sendRequest(oktaAuth, {
    accessToken,
    url: link.href,
    method: link.hints!.allow![0],
    payload,
    transactionClassName
  }));
}
