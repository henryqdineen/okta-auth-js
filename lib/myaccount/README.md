# MyAccount API

## Introduction

MyAccount APIs enables everything needed for the customer's end user to manage one's account. The API assumes the access token is obtained through the OAuth2 either by the classic or IDX flow.

Certain token scopes will be needed to gain the permission to read/manage the account resources:

**profile:**
```
okta.myAccount.profile.read
okta.myAccount.profile.manage
```

**email:**
```
okta.myaccount.email.read
okta.myaccount.email.manage
```

**phone:**
```
okta.myaccount.phone.read
okta.myaccount.phone.manage
```

## How to access API methods

### Via NPM module

MyAccount API is published as a submodule under `@okta/okta-auth-js` package, you can access the myaccount module by following the snippets below:

**ES module:**

```js
import { getEmails } from '@okta/okta-auth-js/myaccount';
```

**CommonJS:**

```js
const { getEmails } = require('@okta/okta-auth-js/myaccount');
```

### Via CDN

The built library bundle is also available on our global CDN. Include the following script in your HTML file to load before your application script:

```html
<script src="https://global.oktacdn.com/okta-auth-js/6.5.0/okta-auth-js.min.js" type="text/javascript"></script>
```

> :warning: The version shown in this sample may be older than the current version. We recommend using the highest version available

Then you can create an instance of the `OktaAuth` object, available globally, then access MyAccount API methods under `myaccount` namespace.

```javascript
const oktaAuth = new OktaAuth({
  // config
});
const emails = await oktaAuth.myaccount.getEmails();
```

## API reference documentation

See [MyAccount API reference documentation](/docs/myaccount/modules/index.md) for detailed API definitions.