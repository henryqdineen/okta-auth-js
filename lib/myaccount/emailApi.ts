import { sendRequest } from './request';
import { 
  IAPIFunction, 
  BaseTransaction, 
  EmailTransaction, 
  EmailChallengeTransaction 
} from './types';

/**
 * @scope: okta.myaccount.email.read
 */
export const getEmails: IAPIFunction<EmailTransaction[]> = async (
  oktaAuth,
  options?
) => {
  const transaction = await sendRequest(oktaAuth, {
    url: '/idp/myaccount/emails',
    method: 'GET',
    accessToken: options?.accessToken,
    transactionClassName: 'EmailTransaction'
  }) as EmailTransaction[];
  return transaction;
};

/**
 * @scope: okta.myaccount.email.read
 */
export const getEmail: IAPIFunction<EmailTransaction> = async (
  oktaAuth, 
  options
) => {
  const { id, accessToken } = options!;
  const transaction = await sendRequest(oktaAuth, {
    url: `/idp/myaccount/emails/${id}`,
    method: 'GET',
    accessToken,
    transactionClassName: 'EmailTransaction'
  }) as EmailTransaction;
  return transaction;
};

/**
 * @scope: okta.myaccount.email.manage
 */
export const addEmail: IAPIFunction<EmailTransaction> = async (
  oktaAuth, 
  options
): Promise<EmailTransaction> => {
  const { accessToken, payload } = options!;
  const transaction = await sendRequest(oktaAuth, {
    url: '/idp/myaccount/emails',
    method: 'POST',
    payload,
    accessToken,
    transactionClassName: 'EmailTransaction'
  }) as EmailTransaction;
  return transaction;
};

/**
 * @scope: okta.myaccount.email.manage
 */
export const deleteEmail: IAPIFunction<BaseTransaction> = async (
  oktaAuth, 
  options
) => {
  const { id, accessToken } = options!;
  const transaction = await sendRequest(oktaAuth, {
    url: `/idp/myaccount/emails/${id}`,
    method: 'DELETE',
    accessToken
  }) as BaseTransaction;
  return transaction;
};

/**
 * @scope: okta.myaccount.email.read
 */
export const sendEmailChallenge: IAPIFunction<EmailChallengeTransaction> = async (
  oktaAuth, 
  options
) => {
  const { id, accessToken } = options!;
  const transaction = await sendRequest(oktaAuth, {
    url: `/idp/myaccount/emails/${id}/challenge`,
    method: 'POST',
    accessToken,
    transactionClassName: 'EmailChallengeTransaction'
  }) as EmailChallengeTransaction;
  return transaction;
};

/**
 * @scope: okta.myaccount.email.read
 */
export const getEmailChallenge: IAPIFunction<EmailChallengeTransaction> = async (
  oktaAuth, 
  options
) => {
  const { emailId, challengeId, accessToken } = options!;
  const transaction = await sendRequest(oktaAuth, {
    url: `/idp/myaccount/emails/${emailId}/challenge/${challengeId}`,
    method: 'POST',
    accessToken,
    transactionClassName: 'EmailChallengeTransaction'
  }) as EmailChallengeTransaction;
  return transaction;
};

/**
 * @scope: okta.myaccount.email.manage
 */
export const verifyEmailChallenge: IAPIFunction<BaseTransaction> = async (
  oktaAuth,
  options
) => {
  const { emailId, challengeId, payload, accessToken } = options!;
  const transaction = await sendRequest(oktaAuth, {
    url: `/idp/myaccount/emails/${emailId}/challenge/${challengeId}/verify`,
    method: 'POST',
    payload,
    accessToken
  }) as BaseTransaction;
  return transaction;
};
