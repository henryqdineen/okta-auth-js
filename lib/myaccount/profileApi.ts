import { sendRequest } from './request';
import { 
  IAPIFunction,
  ProfileTransaction,
  ProfileSchemaTransaction
} from './types';

/**
 * @scope: okta.myaccount.profile.read
 */
export const getProfile: IAPIFunction<ProfileTransaction> = async (oktaAuth, options?) => {
  const transaction = await sendRequest(oktaAuth, {
    url: '/idp/myaccount/profile',
    method: 'GET',
    accessToken: options?.accessToken,
    transactionClassName: 'ProfileTransaction'
  }) as ProfileTransaction;
  return transaction;
};

/**
 * @scope: okta.myaccount.profile.manage
 */
export const updateProfile: IAPIFunction<ProfileTransaction> = async (
  oktaAuth, 
  options
) => {
  const { payload, accessToken } = options!;
  const transaction = await sendRequest(oktaAuth, {
    url: '/idp/myaccount/profile',
    method: 'PUT',
    payload,
    accessToken,
    transactionClassName: 'ProfileTransaction'
  }) as ProfileTransaction;
  return transaction;
};

/**
 * @scope: okta.myaccount.profile.read
 */
export const getProfileSchema: IAPIFunction<ProfileSchemaTransaction> = async (
  oktaAuth, 
  options?
): Promise<ProfileSchemaTransaction> => {
  const transaction = await sendRequest(oktaAuth, {
    url: '/idp/myaccount/profile/schema',
    method: 'GET',
    accessToken: options?.accessToken,
    transactionClassName: 'ProfileSchemaTransaction'
  }) as ProfileSchemaTransaction;
  return transaction;
};
