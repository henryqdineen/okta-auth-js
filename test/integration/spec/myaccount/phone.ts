import {
  addPhone,
  deletePhone,
  sendPhoneChallenge,
  verifyPhoneChallenge
} from '../../../../lib/myaccount';
import {
  createClient,
  signinAndGetTokens
} from '../../util';
import { PhoneTransaction } from '../../../../lib/myaccount/transactions';

// TODO: generate dynamic profile with a18n api
describe('MyAccount Phone API', () => {
  let client, token;

  beforeAll(async () => {
    client = createClient({});
    const {
      tokens: {
        accessToken: { accessToken } = {}
      }
    } = await signinAndGetTokens(client, {
      scopes: [
        'openid',
        'profile',
        'okta.myAccount.phone.read',
        'okta.myAccount.phone.manage'
      ]
    });
    token = accessToken!;
  });

  describe('Manage with Transaction functions', () => {
    let phone: PhoneTransaction;

    afterEach(async () => {
      if (phone) {
        try {
          await phone.delete();
        } catch {
          // do nothing
        }
      }
    });

    it('can manage phone with transaction functions', async () => {
      // create test phone
      phone = await addPhone(client, {
        accessToken: token,
        payload: {
          profile: {
            phoneNumber: '+11234567890'
          },
          // verify link will be included in the response when sendCode is true
          // keep this flag as false in test to avoid sms rate limit issue
          sendCode: false,
          method: 'SMS'
        }
      });
      expect(phone).toMatchSnapshot({
        headers: expect.any(Object),
        _http: {
          headers: expect.any(Object),
        },
        id: expect.any(String),
        profile: {
          phoneNumber: expect.any(String)
        }
      });

      // get phone
      const transaction = await phone.get();
      expect(transaction).toMatchSnapshot({
        headers: expect.any(Object),
        _http: {
          headers: expect.any(Object),
        },
        id: expect.any(String),
        profile: {
          phoneNumber: expect.any(String)
        }
      });

      // send challenge code
      const challenge = await phone.challenge({ method: 'SMS' });
      expect(challenge).toMatchSnapshot({
        _http: {
          headers: expect.any(Object),
        }
      });

      // verify challenge code
      // no verify link when auto send is false
      // use sdk method instead
      // await expect(phone.verify!({ verificationCode: '000000' }))
      //   .rejects
      //   .toThrowErrorMatchingSnapshot();
      await expect(verifyPhoneChallenge(client, {
        accessToken: token,
        id: phone.id,
        payload: { verificationCode: '000000' }
      }))
        .rejects
        .toThrowErrorMatchingSnapshot();

      // delete
      const deleteTransaction = await phone.delete();
      expect(deleteTransaction._http.status).toEqual(204);
    });

  });

  describe('Manage with SDK methods', () => {
    let phone: PhoneTransaction;

    afterEach(async () => {
      if (phone) {
        try {
          await phone.delete();
        } catch {
          // do nothing
        }
      }
    });

    // TODO: integrate with i18n
    it('can manage phone with SDK methods', async () => {
      // create test phone
      phone = await addPhone(client, {
        accessToken: token,
        payload: {
          profile: {
            phoneNumber: '+12345678901'
          },
          sendCode: false,
          method: 'SMS'
        }
      });
      expect(phone).toMatchSnapshot({
        headers: expect.any(Object),
        _http: {
          headers: expect.any(Object),
        },
        id: expect.any(String),
        profile: {
          phoneNumber: expect.any(String)
        }
      });

      // send challenge code
      try {
        const challenge = await sendPhoneChallenge(client, {
          id: phone.id,
          payload: { method: 'SMS' },
          accessToken: token,
        });
        expect(challenge).toMatchSnapshot({
          _http: {
            headers: expect.any(Object),
          }
        });
      } catch (err) {
        // may failure due to - You have reached the limit of sms requests, please try again later.
        console.log(err);
      }

      // verify challenge
      // use invalid code to test the api. expect error
      // TODO: integrate with a18n api to test against real verification code
      await expect(verifyPhoneChallenge(client, {
        accessToken: token,
        id: phone.id,
        payload: { verificationCode: '000000' }
      }))
        .rejects
        .toThrowErrorMatchingSnapshot();

      // delete
      const transaction = await deletePhone(client, {
        accessToken: token,
        id: phone.id
      });
      expect(transaction._http.status).toEqual(204);

    });

  });

});
