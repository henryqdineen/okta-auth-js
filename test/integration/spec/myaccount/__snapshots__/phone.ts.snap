// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyAccount Phone API Manage with SDK methods can manage phone with SDK methods 1`] = `
Object {
  "_http": Object {
    "headers": Any<Object>,
    "status": 201,
  },
  "challenge": [Function],
  "delete": [Function],
  "get": [Function],
  "headers": Any<Object>,
  "id": Any<String>,
  "profile": Object {
    "phoneNumber": Any<String>,
  },
  "status": "UNVERIFIED",
}
`;

exports[`MyAccount Phone API Manage with SDK methods can manage phone with SDK methods 2`] = `
Object {
  "_http": Object {
    "headers": Any<Object>,
    "status": 200,
  },
}
`;

exports[`MyAccount Phone API Manage with SDK methods can manage phone with SDK methods 3`] = `"Api validation failed: factorId and/or verificationCode"`;

exports[`MyAccount Phone API Manage with Transaction functions can manage phone with transaction functions 1`] = `
Object {
  "_http": Object {
    "headers": Any<Object>,
    "status": 201,
  },
  "challenge": [Function],
  "delete": [Function],
  "get": [Function],
  "headers": Any<Object>,
  "id": Any<String>,
  "profile": Object {
    "phoneNumber": Any<String>,
  },
  "status": "UNVERIFIED",
}
`;

exports[`MyAccount Phone API Manage with Transaction functions can manage phone with transaction functions 2`] = `
Object {
  "_http": Object {
    "headers": Any<Object>,
    "status": 200,
  },
  "challenge": [Function],
  "delete": [Function],
  "get": [Function],
  "headers": Any<Object>,
  "id": Any<String>,
  "profile": Object {
    "phoneNumber": Any<String>,
  },
  "status": "UNVERIFIED",
}
`;

exports[`MyAccount Phone API Manage with Transaction functions can manage phone with transaction functions 3`] = `
Object {
  "_http": Object {
    "headers": Any<Object>,
    "status": 200,
  },
}
`;

exports[`MyAccount Phone API Manage with Transaction functions can manage phone with transaction functions 4`] = `"Api validation failed: factorId and/or verificationCode"`;
